iceoryx_hoofs/include/iceoryx_hoofs/design_pattern/polymorphic_handler.hpp

```diff
@@ -0,0 +1,126 @@
+struct DefaultHooks
+{
+    /// @brief called if the polymorphic handler is set or reset after finalize
+    /// @param currentInstance the current instance of the handler singleton
+    /// @param newInstance the instance of the handler singleton to be set
+    static void onSetAfterFinalize(Interface& currentInstance, Interface& newInstance) noexcept;
+};
+
+} // namespace detail
+
+/// @brief Implements a singleton handler that has a default instance and can be changed
+/// to another instance at runtime. All instances have to derive from the same interface.
+/// The singleton handler owns the default instance but all other instances are created externally.
+/// @tparam Interface The interface of the handler instances.
+/// @tparam Default The type of the default instance. Must be equal to or derive from Interface.
+/// @tparam Hooks A struct that implements onSetAfterFinalize which is called when
+/// attempting to set or reset the handler after finalize was called.
+///
+/// @note In the special case where Default equals Interface, no polymorphism is required.
+/// It is then possible to e.g. switch between multiple instances of Default type.
+/// @note The lifetime of external non-default instances must exceed the lifetime of the PolymorphicHandler.
+/// @note The PolymorphicHandler is guaranteed to provide a valid handler during the whole program lifetime (static).
+/// It is hence not advisable to have other static variables depend on the PolymorphicHandler.
+/// It must be ensured that they are destroyed before the PolymorphicHandler.
+/// @note Hooks must implement
+/// static void onSetAfterFinalize(Interface& /*currentInstance*/, Interface& /*newInstance*/).
+template <typename Interface, typename Default, typename Hooks = detail::DefaultHooks<Interface>>
+class PolymorphicHandler
+{
+    static_assert(std::is_base_of<Interface, Default>::value, "Default must inherit from Interface");
+
+    using Self = PolymorphicHandler<Interface, Default, Hooks>;
+    friend class StaticLifetimeGuard<Self>;
+
+  public:
+    /// @brief set the current singleton instance
+    /// @return the current instance
+    /// @note we cannot be sure to use the current handler unless we call get,
+    /// i.e. a reference obtained from get may reference a previous handler
+    /// (that is still functional but inactive)
+    static Interface& get() noexcept;
+
+    /// @brief set the current singleton instance
+    /// @param handlerGuard a guard to the handler instance to be set
+    /// @return pointer to the previous instance or nullptr if the handler could not be replaced
+    /// @note the handler cannot be replaced if it was finalized
+    /// @note using a guard in the interface prevents the handler to be destroyed while it is used,
+    ///       passing the guard by value is necessary (it has no state anyway)
+    template <typename Handler>
+    static Interface* set(StaticLifetimeGuard<Handler> handlerGuard) noexcept;
```

Comment:
Yes, to restore the previous handler easily in a concurrent setting. Getting it before and setting it will not guarantee that. A return value is also needed to indicate failure due to finalization (should be the only reason for failure).

Reply:
It might, have to think about it. The intention of providing the guard as argument is to ensure the handler to be set is a singleton and still alive.

So the same guarantee should be possible for the return value but the problem is: it could be nothing to return if setting fails and then it would need to be an optional etc. which is really ugly (do not want a pointer to the guard, where would the guard live? - the instance does not have this problem as it is static).

Reply:
@elBoberido Changed `set` and `reset` to return bool to indicate failure. There is no way to use a `StaticLifetimeGuard<Interface>` of a polymorphic interface, hence there was no option to return such an object.

As it is not too useful anyway, the return type was simplified to `bool`.